processing/src/main/java/io/druid/query/groupby/epinephelinae/GroupByMergingQueryRunnerV2.java

```diff
@@ -332,6 +313,39 @@ public void cleanup(CloseableGrouperIterator<RowBasedKey, Row> iterFromMake)
     );
   }
 
+  private List<ReferenceCountingResourceHolder<ByteBuffer>> getMergeBuffersHolder(
+      int numBuffers,
+      boolean hasTimeout,
+      long timeoutAt
+  )
+  {
+    try {
+      if (numBuffers > mergeBufferPool.maxSize()) {
+        throw new ResourceLimitExceededException(
+            "Query needs " + numBuffers + " merge buffers, but only "
+            + mergeBufferPool.maxSize() + " merge buffers were configured"
+        );
+      }
+      final List<ReferenceCountingResourceHolder<ByteBuffer>> mergeBufferHolder;
+      // This will potentially block if there are no merge buffers left in the pool.
+      if (hasTimeout) {
+        final long timeout = timeoutAt - System.currentTimeMillis();
+        if (timeout <= 0) {
+          throw new TimeoutException();
+        }
+        if ((mergeBufferHolder = mergeBufferPool.takeBatch(numBuffers, timeout)).isEmpty()) {
+          throw new InsufficientResourcesException("Cannot acquire enough merge buffers");
```

Comment:
Maybe this should be a `TimeoutException`? (One of the 4 types that gives callers a special error code; see QueryInterruptedException's `getErrorCodeFromThrowable` method)

The current wording and type does not make it clear that this is really a timeout error.

Reply:
Good point. Changed exception.